ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForMajor.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hive.metastore.ColumnType;
+import org.apache.hadoop.hive.metastore.api.CompactionType;
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+import org.apache.hive.common.util.HiveStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Builds query strings that help with query-based MAJOR compaction of CRUD.
+ */
+class CompactionQueryBuilderForMajor extends CompactionQueryBuilder {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CompactionQueryBuilderForMajor.class.getName());
+
+  /**
+   * Construct a CompactionQueryBuilderForMajor with required params.
+   *
+   * @param compactionType major or minor or rebalance, e.g. CompactionType.MAJOR.
+   *                       Cannot be null.
+   * @param operation query's Operation e.g. Operation.CREATE.
+   * @throws IllegalArgumentException if compactionType is null
+   */
+  CompactionQueryBuilderForMajor(CompactionType compactionType, Operation operation, String resultTableName) {
+    super(compactionType, operation, false, resultTableName);
+  }
```

Comment:
`CompactionType.Major` could be passed directly into super call (hardcoded) instead of taking it as a parameter. Passing other values can cause this class to output incorrect SQL. The same applies for the other implementations.